src/mcpwm/timer_connection.rs

```diff
@@ -0,0 +1,158 @@
+use crate::mcpwm::Group;
+
+use super::{
+    comparator::OptionalCmpCfg,
+    generator::OptionalGenCfg,
+    operator::{self, NoOperator, OptionalOperator, OPERATOR},
+    timer::Timer,
+    Operator, OperatorConfig,
+};
+
+// TODO: How do we want fault module to fit into this?
+/// Created by `Timer::into_connection()`
+pub struct TimerConnection<const N: u8, G: Group, O0, O1, O2>
+where
+    O0: OptionalOperator<0, G>,
+    O1: OptionalOperator<1, G>,
+    O2: OptionalOperator<2, G>,
+{
+    timer: Timer<N, G>,
+    operator0: O0,
+    operator1: O1,
+    operator2: O2,
+}
+
+impl<const N: u8, G: Group> TimerConnection<N, G, NoOperator, NoOperator, NoOperator> {
+    pub(crate) fn new(timer: Timer<N, G>) -> Self {
+        Self {
+            timer,
+            operator0: NoOperator,
+            operator1: NoOperator,
+            operator2: NoOperator,
+        }
+    }
+}
+
+// Since there can only ever by one instance of every operator type (except NoOperator)
+// we know that there can be no mem::swap or similar to cause any problems.
+//
+// Thus we know that after split is called nothing can be added/removed while still having access to
+// the individual objects. We also garantuee that the operators wont live longer than the timer
+impl<
+        const N: u8,
+        G: Group,
+        O0: OptionalOperator<0, G>,
+        O1: OptionalOperator<1, G>,
+        O2: OptionalOperator<2, G>,
+    > TimerConnection<N, G, O0, O1, O2>
+{
+    pub fn split(&mut self) -> (&mut Timer<N, G>, &mut O0, &mut O1, &mut O2) {
+        (
+            &mut self.timer,
+            &mut self.operator0,
+            &mut self.operator1,
+            &mut self.operator2,
+        )
+    }
+}
+// TODO: Do something more builder-pattern like for making the operator?
+impl<const N: u8, G, O1, O2> TimerConnection<N, G, NoOperator, O1, O2>
+where
+    G: Group,
+    O1: OptionalOperator<1, G>,
+    O2: OptionalOperator<2, G>,
+{
+    pub fn attatch_operator0<CMPX, CMPY, GENA, GENB>(
```

Comment:
So for v4.4, these bounds don't currently stop you from using timer0 & operator2, right? Perhaps we can somehow use const generics here such that TimerConnection<N> must always match Operator<N> for v4.4 :thinking:.

Reply:
Something roughly like(semi made up code with bad names)?

```rust
struct CanAttach<const TIMER: u8, const OPERATOR: u8>;
trait Yes;

#[cfg(v4_4)]
impl Yes for CanAttach<0, 0>{}

#[cfg(v4_4)]
impl Yes for CanAttach<1, 1>{}

#[cfg(v4_4)]
impl Yes for CanAttach<2, 2>{}

#[cfg(not(v4_4)]
impl<const TIMER: u8, const OPERATOR: u8> Yes for CanAttach<TIMER, OPERATOR>{}
```

```diff
 impl<const N: u8, G, O1, O2> TimerConnection<N, G, NoOperator, O1, O2>
 where
     G: Group,
     O1: OptionalOperator<1, G>,
     O2: OptionalOperator<2, G>,
+    CanAttach<N, 0>: Yes,
 {
     pub fn attatch_operator0<CMPX, CMPY, GENA, GENB>(
```

Reply:
Yep! This is basically what I had in mind - maybe one day when const expressions are stable we can simply further.